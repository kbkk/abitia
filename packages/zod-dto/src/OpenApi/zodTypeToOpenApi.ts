import {
    ZodArray, ZodBigInt, ZodBoolean, ZodDefault,
    ZodEnum,
    ZodLiteral,
    ZodNullable, ZodNumber,
    ZodObject,
    ZodOptional, ZodString,
    ZodTransformer, ZodTuple,
    ZodTypeAny,
    ZodTypeDef,
    ZodUnion,
} from 'zod';

import { OpenApiBuilderProperties } from './types';

export interface OpenApiElement {
    [key: string]: unknown;
    type: string;
}

export interface ZodTypeDefOpenApi extends ZodTypeDef {
    openApi: OpenApiBuilderProperties;
}

export const zodTypeToOpenApi = (zodType: ZodTypeAny): OpenApiElement => {
    const zodDef = zodType._def;

    const openApiElement = (element: OpenApiElement): OpenApiElement => {
        return {
            required: true,
            ...element,
            ...(zodDef as ZodTypeDefOpenApi).openApi,
        };
    };

    switch(zodType.constructor.name) {
    case ZodObject.name: {
        const shape = zodDef.shape();
        const shapeKeys = Object.keys(shape);
        const properties = {};

        for (const key of shapeKeys) {
            const propZodType = shape[key];

            properties[key] = zodTypeToOpenApi(propZodType);
        }

        return openApiElement({
            type: 'object',
            properties,
        });
    }
    case ZodDefault.name:
    case ZodOptional.name: {
        return openApiElement({
            ...zodTypeToOpenApi(zodDef.innerType),
            required: false,
        });
    }
    case ZodNullable.name: {
        return openApiElement({
            ...zodTypeToOpenApi(zodDef.innerType),
            nullable: true,
        });
    }
    case ZodTransformer.name: {
        return openApiElement({
            ...zodTypeToOpenApi(zodDef.schema),
        });
    }
    case ZodEnum.name: {
        const enumValues = zodDef.values;

        return openApiElement({
            type: 'string',
            enum: enumValues,
        });
    }
    case ZodLiteral.name: {
        const { value } = zodDef;

        return openApiElement({
            type: 'string',
            enum: [value],
        });
    }
    case ZodUnion.name: {
        const { options } = zodDef;

        return openApiElement({
            oneOf: options.map(item => zodTypeToOpenApi(item)),
        } as unknown as OpenApiElement);
    }
    case ZodTuple.name: {
        // Switch to OpenAPI 3.1 once supported by swagger ui https://stackoverflow.com/questions/57464633/

        const { items } = zodDef;

        return openApiElement({
            type: 'array',
            items: {
                oneOf: items.map(item => zodTypeToOpenApi(item)),
            },
            minItems: items.length,
            maxItems: items.length,
        });
    }
    case ZodArray.name: {
        const { type } = zodDef;

        const items = type ? zodTypeToOpenApi(type) : {};

        return openApiElement({
            type: 'array',
            items,
        });
    }
    case ZodBigInt.name: {
        return openApiElement({
            type: 'integer',
            format: 'int64',
        });
    }
    case ZodString.name: {
        return openApiElement({
            type: 'string',
        });
    }
    case ZodNumber.name: {
        return openApiElement({
            type: 'number',
        });
    }
    case ZodBoolean.name: {
        return openApiElement({
            type: 'boolean',
        });
    }
    default: {
        return openApiElement({
            type: 'string',
        });
    }
    }
};
