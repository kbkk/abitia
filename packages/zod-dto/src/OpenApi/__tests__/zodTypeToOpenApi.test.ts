import * as z from 'zod';
import { ZodTypeAny } from 'zod';

import { zodTypeToOpenApi } from '../zodTypeToOpenApi';

const testTuple = z.tuple([
    z.string(),
    z.object({ name: z.literal('Rudy') }),
    z.array(
        z.union([
            z.literal('blue'),
            z.literal('red'),
        ]),
    )]);

const complexTestSchema = z.object({
    item: z.string().min(5),
    price: z.number(),
    type: z.enum(['buy-it-now', 'auctions'])
        .default('buy-it-now'),
    testTuple,
    kekity: z.union([z.literal('kek'), z.literal('topkek')]),
    arr: z.array(z.string()).max(3),
    obj: z.object({
        uuid: z.string().uuid(),
    }),
});

it('should serialize a complex schema', () => {
    const openApiObject = zodTypeToOpenApi(complexTestSchema);

    expect(openApiObject).toMatchSnapshot();
});

it('should serialize objects', () => {
    const schema = z.object({
        prop1: z.string(),
    });
    const openApiObject = zodTypeToOpenApi(schema);

    expect(openApiObject).toEqual({
        type: 'object',
        required: true,
        properties:  {
            prop1:  {
                type: 'string',
                required: true,
            },
        },
    });
});

it('should serialize nullable types', () => {
    const schema = z.string().nullable();
    const openApiObject = zodTypeToOpenApi(schema);

    expect(openApiObject).toEqual({
        type: 'string',
        required: true,
        nullable: true,
    });
});

it('should serialize optional types', () => {
    const schema = z.string().optional();
    const openApiObject = zodTypeToOpenApi(schema);

    expect(openApiObject).toEqual({
        type: 'string',
        required: false,
    });
});

it('should serialize types with default value', () => {
    const schema = z.string().default('abitia');
    const openApiObject = zodTypeToOpenApi(schema);

    expect(openApiObject).toEqual({
        type: 'string',
        required: false,
    });
});

it('should serialize enums', () => {
    const schema = z.enum(['adama', 'kota']);
    const openApiObject = zodTypeToOpenApi(schema);

    expect(openApiObject).toEqual({
        type: 'string',
        required: true,
        enum: ['adama', 'kota'],
    });
});

describe('scalar types', () => {
    const testCases: [ZodTypeAny, string, string?][] = [
        // [zod type, expected open api type, expected format]
        [z.string(), 'string'],
        [z.number(), 'number'],
        [z.boolean(), 'boolean'],
        [z.bigint(), 'integer', 'int64'],
        // [z.null(), 'null'], <- Needs OpenApi 3.1 to be represented correctly
        // [z.undefined(), 'undefined'], <- TBD, probably the property should be removed from schema
    ];

    for(const [zodType, expectedType, expectedFormat] of testCases) {
        it(expectedType, () => {
            const openApiObject = zodTypeToOpenApi(zodType);

            expect(openApiObject).toEqual({
                type: expectedType,
                required: true,
                format: expectedFormat ?? undefined,
            });
        });
    }
});
