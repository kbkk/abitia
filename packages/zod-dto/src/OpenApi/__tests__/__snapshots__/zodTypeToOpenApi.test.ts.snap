// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should serialize a complex schema 1`] = `
Object {
  "properties": Object {
    "arr": Object {
      "items": Object {
        "required": true,
        "type": "string",
      },
      "required": true,
      "type": "array",
    },
    "item": Object {
      "required": true,
      "type": "string",
    },
    "kekity": Object {
      "oneOf": Array [
        Object {
          "enum": Array [
            "kek",
          ],
          "required": true,
          "type": "string",
        },
        Object {
          "enum": Array [
            "topkek",
          ],
          "required": true,
          "type": "string",
        },
      ],
      "required": true,
    },
    "obj": Object {
      "properties": Object {
        "uuid": Object {
          "required": true,
          "type": "string",
        },
      },
      "required": true,
      "type": "object",
    },
    "price": Object {
      "required": true,
      "type": "number",
    },
    "testTuple": Object {
      "items": Object {
        "oneOf": Array [
          Object {
            "required": true,
            "type": "string",
          },
          Object {
            "properties": Object {
              "name": Object {
                "enum": Array [
                  "Rudy",
                ],
                "required": true,
                "type": "string",
              },
            },
            "required": true,
            "type": "object",
          },
          Object {
            "items": Object {
              "oneOf": Array [
                Object {
                  "enum": Array [
                    "blue",
                  ],
                  "required": true,
                  "type": "string",
                },
                Object {
                  "enum": Array [
                    "red",
                  ],
                  "required": true,
                  "type": "string",
                },
              ],
              "required": true,
            },
            "required": true,
            "type": "array",
          },
        ],
      },
      "maxItems": 3,
      "minItems": 3,
      "required": true,
      "type": "array",
    },
    "type": Object {
      "enum": Array [
        "buy-it-now",
        "auctions",
      ],
      "required": false,
      "type": "string",
    },
  },
  "required": true,
  "type": "object",
}
`;
